app/controllers/v0/sessions_controller.rb

```diff
@@ -112,13 +103,30 @@ def set_cookies
       set_sso_cookie! # Sets a cookie "vagov_session_<env>" with attributes needed for SSO.
     end
 
+    def new_query_params
+      query_params = {}
+      query_params[:clientId] = params[:clientId] if params[:clientId]
+      query_params[:op] = 'signup' if params[:type] == 'signup'
+      query_params[:RelayState] = { type: params[:type] }
+      query_params
+    end
+
+    def login_redirect_query_params(auth, code, relay_state)
+      query_params = {}
+      query_params[:auth] = 'fail' if auth == 'fail'
+      query_params[:code] = code if code.present?
+      query_params[:type] = relay_state['type'] if relay_state['type']
+      query_params
+    end
```

Comment:
ooooof, is all of this necessary? Why not just pass all of the controller params into url_service. You don't need controller and path and action params, but you can figure out the ones you need there in URLService. 

I do like the idea of the login_redirect_url being entirely generated in url_service though, so that makes sense and would help clean up session controller a bit, but I'm thinking if you are creating a separate distinct path for signup, that doesn't require a query param -- then type is really all you need, and the rest of this can live and be built entirely in url_service, no?

Reply:
I'm not sure URLService should be responsible for picking what URL params it wants depending on which URL method is invoked. It gets more tangled when you consider that the redirect URLs currently directly use all of the query params that are passed to the instance. And to avoid that, it might be messy to choose what specific params each method should use.

Either way, this picking and choosing and query transformation has to live somewhere. My thinking was that URL service should be flexible with receiving query params, and the only special handling it does might be for RelayState, but I kept that deliberately minimal too.

I was going back and forth for the `op=signup`. I wasn't sure about URLService conditionally adding its own query params, but it does something like that already for RelayState, so probably not a big deal. Maybe I can re-introduce the `signup_url` method where the only extra thing it does is set `@query_params[:op] = 'signup'`.

Reply:
it's not picking, its based entirely on the route, in particular the params for "type" cannot be overridden by user input, it's based entirely on the supported routes to the controller. And op=signup is just something that gets added when route == signup, but not for any of the other routes. What you're doing right now with query_params is essentially the same, you're normalizing them but I don't think you need to. 

The only routes that will ever care or concern themselves with the RelayState are the callback routes. RelayState is entirely constructed in URLService for that reason. The FE should not be using RelayState either really, it's strictly a backend thing so thats why I don't understand why you would want to construct RelayState outside of URLService, it should only relevant to given route / url for now. Later when we introduce SessionActivity, we'll use those RelayState attributes in the controller, but that will just parsing existing ones, not updating future ones.

Reply:
Wouldn't we have to specify what params to keep if we send all of them from the controller to the service? We'd be normalizing the query but in URLService instead. Right now, for any `new` actions, we only care about `clientId` and `type`. For `op=signup`, I can understand moving it to the service, because that's just determined based on the type being passed. The `query_params` being passed into URLService were meant to be directly appended to the base URLs that we construct. Like we probably don't want to pass anything other than `clientId` and `type` (in the RelayState, specifically) to the URL from `build_sso_url`. So as the effect of passing the unfiltered query params to `URLService`, if we have to limit it so that we don't pass extraneous queries to that URL, we have to specify that we want those params.

Reply:
I'm also making the distinction between having `type` in the query and `type` in the RelayState in the query. In order for URLService to know to merge `type` into the RelayState, it needs to explicitly expect that in the `query_params`.

Reply:
Do you think it would be better to do something like this?

```
def idme_url
  @query_params[:RelayState] = relay_state_params(type: 'idme')
  build_sso_url(LOA::IDME_LOA1)
end

def signup_url
  @query_params[:op] = 'signup'
  @query_params[:RelayState] = relay_state_params(type: 'signup')
  idme_url
end

...

def relay_state_params(extra_params = {})
  { originating_request_id: Thread.current['request_id'] }.merge(extra_params).to_json
end
```

Reply:
I originally made RelayState modifiable from outside of URLService for the possibility of directly passing some other query from the request. But maybe that's not likely to happen and we don't need to account for that right now. This still leaves `clientId`, which I think should be explicitly passed to `URLService` from the `new` action rather than the service figuring out if it does have that param.

Reply:
well if signup is a route you won't need to do that, i guess you would in the interim until the FE can change things, but yea, that looks a lot cleaner to me. And if you're really concerned with passing too many params to URLService you could always do:

`query_params: params.slice(:client_id, :type)`

Reply:
I think I actually favor having all of the arguments to URLService as named arguments, it's much cleaner and easier to assign to instance variables and actually, I'm pretty sure we're going to need `originating_request_id` to be an argument eventually too. The reason is that when we automatically LOA3 a user who has previously identity proofed, the request_id is going to be different than the originating_request_id we have in RelayState, so its going to be up to SessionsController to pass that as an argument.

Reply:
if type is defined as an instance variable at initialization, you can just do all of the relay state changes in build_sso_url method.